docs/source/DeveloperGuide/Python3Migration/index.rst

```diff
@@ -0,0 +1,227 @@
+This quickly turned out to be impractical for anything other than small projects and this is especially true of DIRAC where a large fraction of the code is not tested automatically due to it depending on external services.
+Running ``2to3`` at install time isn't ideal as it make it hard to map line numbers to the code, introduces new bugs and generally makes "ugly" code.
+Instead the strategy used by almost every project has been to move to a code base which is compatible with both Python 2 and Python 3 at the same time.
+This is not inherently an additional burden as the "modern" style of Python 2 code is compatible with Python 3 so it is beneficial when using both Python versions.
+
+Further reading
+---------------
+
+The following links contain useful information about migrating to Python 3:
+
+ - `The Conservative Python 3 Porting Guide <https://portingguide.readthedocs.io/en/latest/index.html>`_
+ - `Mercurial's Journey to and Reflections on Python 3 <https://gregoryszorc.com/blog/2020/01/13/mercurial%27s-journey-to-and-reflections-on-python-3/>`_
+
+
+Recommendations for code
+------------------------
+
+The "The Conservative Python 3 Porting Guide" linked above is an excellent source of information.
+This sections contains some details that are particularly relevant to DIRAC.
+
+**bytes vs str**
+  The most difficult change when moving Python 3 is the splitting of the `str` type one for text and one for true binary data.
+  This exposes subtle issues in Python 2 that were likely never noticed and an automatic conversion to fix this is inherently impossible.
+  More details about this can be found `here <https://portingguide.readthedocs.io/en/latest/strings.html>`_.
+
+  In most situations DIRAC is only dealing with ascii or unicode strings and therefore nothing needs to change.
+  However many libraries choose to be independent of the character encoding used and therefore return a ``bytes`` object in Python 3 instead of ``str``
+
+  .. code-block:: python
+
+    result = subprocess.check_output(["echo", "Hello"])
+
+    # Bad: Fails on Python 3 with "TypeError: can't concat str to bytes"
+    return "Result is" + result
+
+    # Good: Explicitly decode bytes to str (does nothing on Python 2)
+    return "Result is" + result.decode()
+
+    # For subprocess functions, the universal_newlines=True argument can be used
+    other_result = subprocess.check_output(["echo", "Hello"], universal_newlines=True)
+    # Good: other_result is already a str object
+    return "Result is" + other_result
+
+  **Checking the type of a string:**
+
+  .. code-block:: python
+
+    # Bad: Types should be check using isinstance
+    if type(my_variable) == str:
+
+    # Bad: basestring does not exist in Python 3
+    if instance(my_variable, basestring):
+
+    # Good: Supports both Python 2 and 3
+    if instance(my_variable, six.string_types):
```

Comment:
It doesn't:

```
In [1]: a = u'hello world'
In [2]: isinstance(a, str)
Out[2]: False
```

Reply:
Yes, in python2 does not work in some cases. But when we know a is not unicode we should not use six.string_types. No? I had no problem to make Graph package work with python2 and python3 https://github.com/DIRACGrid/DIRAC/pull/4497

Reply:
I think the recommendation should be to use `six.string_types` in all situations to avoid any possible mistakes, especially as it doesn't have an obvious failure mode.

I don't see any downside and it's easy to refactor away once we drop Python 2 for good.